plugins/inputs/win_wmi/win_wmi.go

```diff
@@ -0,0 +1,265 @@
+		return fmt.Errorf("failed to QueryInterface: %v", err)
+	}
+	defer wmi.Release()
+
+	// service is a SWbemServices
+	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer", nil, q.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed calling method ConnectServer: %v", err)
+	}
+	service := serviceRaw.ToIDispatch()
+	defer func() { _ = serviceRaw.Clear() }()
+
+	// result is a SWBemObjectSet
+	resultRaw, err := oleutil.CallMethod(service, "ExecQuery", q.query)
+	if err != nil {
+		return fmt.Errorf("failed calling method ExecQuery for query %s: %v", q.query, err)
+	}
+	result := resultRaw.ToIDispatch()
+	defer func() { _ = resultRaw.Clear() }()
+
+	count, err := oleInt64(result, "Count")
+	if err != nil {
+		return fmt.Errorf("failed getting Count: %v", err)
+	}
+
+	for i := int64(0); i < count; i++ {
+		// item is a SWbemObject
+		itemRaw, err := oleutil.CallMethod(result, "ItemIndex", i)
+		if err != nil {
+			return fmt.Errorf("failed calling method ItemIndex: %v", err)
+		}
+
+		outerErr := func() error {
+			item := itemRaw.ToIDispatch()
+			defer item.Release()
+
+			for _, wmiProperty := range q.Properties {
+				innerErr := func() error {
+					prop, err := oleutil.GetProperty(item, wmiProperty)
+					if err != nil {
+						return fmt.Errorf("failed GetProperty: %v", err)
+					}
+					defer func() { _ = prop.Clear() }()
+
+					if q.tagFilter.Match(wmiProperty) {
+						valStr, err := internal.ToString(prop.Value())
+						if err != nil {
+							return fmt.Errorf("converting property %q failed: %v", wmiProperty, err)
+						}
+						tags[wmiProperty] = valStr
+					} else {
+						var fieldValue interface{}
+						switch v := prop.Value().(type) {
+						case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64:
+							fieldValue = v
+						case string:
+							// still might be an int because WMI
+							valInt, err := strconv.ParseInt(v, 10, 64)
+							if err == nil {
+								fieldValue = valInt
+							} else {
+								fieldValue = v
+							}
+						case bool:
+							fieldValue = v
+						case []byte:
+							fieldValue = string(v)
+						case fmt.Stringer:
+							fieldValue = v.String()
+						case nil:
+							fieldValue = nil
+						default:
+							return fmt.Errorf("property %q of type \"%T\" unsupported", wmiProperty, v)
+						}
+						fields[wmiProperty] = fieldValue
+					}
+
+					return nil
+				}()
+
+				if innerErr != nil {
+					return innerErr
+				}
+			}
+			acc.AddFields(q.ClassName, fields, tags)
+			return nil
+		}()
+
+		if outerErr != nil {
+			return outerErr
+		}
+	}
+	return nil
+}
+
+// Gather function
+func (s *Wmi) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
+	for _, query := range s.Queries {
+		wg.Add(1)
+		go func(q Query) {
+			defer wg.Done()
+			start := time.Now()
+			err := q.DoQuery(acc)
+			if err != nil {
+				acc.AddError(err)
+			}
+			elapsed := time.Since(start)
+			s.Log.Debugf("Query \"%s\" took %s", q.query, elapsed)
```

Comment:
Not sure if this level of debug should be around for every gather. Is this truly useful?

Reply:
I ripped out the debugging statement.